cookbooks/tian_parameterization_kinematic_slab/coupled-two-phase-tian-parameterization-kinematic-slab.prm

```diff
@@ -0,0 +1,295 @@
+  end
+# Set a stricter linear solver tolerance to ensure convergence
+  subsection Stokes solver parameters
+    set Linear solver tolerance             = 1e-10
+    set Number of cheap Stokes solver steps = 1000
+  end
+end
+
+# Initialize the compositional fields. When using the tian approximation
+# Fluid-solid reaction scheme in the Reactive Fluid Transport material
+# model, the fields 'porosity', 'bound_fluid', 'peridotite', 'gabbro',
+# 'MORB', and 'sediment' are all required compositional fields. This is
+# because they are all treated as special fields by the material model.
+# The 4 lithologies ('sediment', 'MORB', 'gabbro', 'peridotite') are used
+# to determine the equilibrium wt% of water that can be incorporated into
+# the corresponding lithology at the model P-T conditions. Free fluid
+# ('porosity') is then partitioned into/out of the solid depending on the
+# current wt% water within the lithology ('bound_fluid'), and how much 'porosity'
+# is present.
+subsection Compositional fields
+  set Number of fields = 6
+  set Names of fields = porosity, bound_fluid, peridotite, gabbro, MORB, sediment
+  set Compositional field methods = field, field, field, field, field, field
+end
+
+# Initially, the interior of the model is just the periodite mantle.
+# The composition evolves through the composition boundary conditions
+# specified below.
+subsection Initial composition model
+  set Model name = function
+  subsection Function
+    set Function expression = 0; 0; 1; 0; 0; 0
+  end
+end
+
+# The slab begins at the left most boundary, at first consisting solely of dry
+# lithologic layers. After 100 kyr, the incoming sediment is hydrated to 3 wt% water,
+# MORB to 2 wt% water, and gabbro to 1 wt% water.
+subsection Boundary composition model
+  set Fixed composition boundary indicators         = left
+  set List of model names                           = function
+  subsection Function
+    set Function constants  = initial_porosity=0, initial_bound_sed=0.03, initial_bound_MORB=0.02, initial_bound_gabbro=0.01, \
+                              sediment_min=0, sediment_max=5e3, MORB_min=5e3, MORB_max=12e3, gabbro_min=12e3, gabbro_max=20e3, \
+                              onset_time=100e3, slab_thickness=20e3
+    set Function expression = initial_porosity; \
+                              if( t<=onset_time, 0, \
+                              if( (y<=slab_thickness - sediment_min) & (y>slab_thickness - sediment_max), initial_bound_sed, \
+                              if( (y<=slab_thickness - MORB_min) & (y>slab_thickness - MORB_max), initial_bound_MORB, \
+                              if( (y<=slab_thickness - gabbro_min) & (y>slab_thickness - gabbro_max), initial_bound_gabbro, 0.0)))); \
+                              if(y>slab_thickness, 1, 0); \
+                              if(y<=8e3, 1, 0); \
+                              if(y>8e3 & y<=15e3, 1, 0); \
+                              if(y>15e3 & y<=slab_thickness, 1, 0)
```

Comment:
Some reformatting suggestions to make the function expression a bit easier to read:

```
    set Function expression = initial_porosity; \
                              if( t<=onset_time, 0, \
                                  if( (y<=slab_thickness - sediment_min) & (y>slab_thickness - sediment_max), initial_bound_sed, \
                                      if( (y<=slab_thickness - MORB_min) & (y>slab_thickness - MORB_max), initial_bound_MORB, \
                                          if( (y<=slab_thickness - gabbro_min) & (y>slab_thickness - gabbro_max), initial_bound_gabbro, 0.0) \
                                        ) \
                                    ) \
                                ); \
                              if(y>slab_thickness, 1, 0); \
                              if(y<=8e3, 1, 0); \
                              if(y>8e3 & y<=15e3, 1, 0); \
                              if(y>15e3 & y<=slab_thickness, 1, 0)
```